### DISCUSSION TITLE
I wanna know how to debug memtest86+

### DISCUSSION THREAD

**Discussion by chlo2ophyll** [memtest86plus/memtest86plus#459] [Show and tell]
![Snipaste_2024-10-02_17-02-27](https://github.com/user-attachments/assets/110bb69c-8375-4717-8be5-d3738b04d74c)
I don't know how to do this

**Comment by chlo2ophyll** [comment]
I am a novice. I want to learn how to debug step by step to get familiar with the entire source code process. Who can teach me? Thank you very much.

**Comment by debrouxl** [comment]
Your screenshot shows that you already seem to have launched QEMU through `debug_memtest.sh`, so you're nearly done :)
From the EFI shell, launch
```
fs0:
memtest.efi
```
Out of the box, the debugger will break onto `main` and `run_at`. The rest is standard GDB usage: `cont`, `disable`, `break`, `step`, `stepi`, `disas`, `x/... address` are the commands I use most often, I'm not remotely an expert of GDB.

**Comment by chlo2ophyll** [comment]
wow wow thank you so fucking much!

**Comment by chlo2ophyll** [comment]
![Snipaste_2024-10-03_10-53-36](https://github.com/user-attachments/assets/cd527933-306c-4050-95a6-a54bd49b841f)
I have a new question. I have changed start_run to true. Why is it still false when it is read?

**Comment by debrouxl** [comment]
The debug information is loaded at two addresses, as you can see in your screenshot: it's because the program periodically relocates itself, in order to be able to test almost the entire available memory range. However, the wrong address is usually chosen for operations such as `p <variable_name>`...
Fix welcome, if you know how to improve this :)

**Comment by chlo2ophyll** [comment]
I'm really sorry, I don't have this ability yet, so can I fix the area to be tested so that the variable shows the real and correct value during debugging? How can I fix an address in the code for testing? Please give me some advice
